File: 654-最大二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> TreeNode:
        return self.construct(nums, 0, len(nums) - 1)

    def construct(self, nums, left, right):
        if left > right:
            return None

        max_index = left
        for i in range(left, right + 1):
            if nums[i] > nums[max_index]:
                max_index = i
        root = TreeNode(nums[max_index])
        root.left = self.construct(nums, left, max_index - 1)
        root.right = self.construct(nums, max_index + 1, right)

        return root

File: 654-最大二叉树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums);
    TreeNode* construct(vector<int>& nums, int left, int right);
};

TreeNode* Solution::constructMaximumBinaryTree(vector<int>& nums){
    return construct(nums, 0, nums.size() - 1);
}

TreeNode* Solution::construct(vector<int>& nums, int left, int right){
    if(left > right)
        return NULL;

    int max_index = left;
    for(int i = left; i <= right; i++){
        if(nums[i] > nums[max_index])
            max_index = i;
    }
    TreeNode* root = new TreeNode(nums[max_index]);
    root->left = construct(nums, left, max_index - 1);
    root->right = construct(nums, max_index + 1, right);

    return root;
}
